src/TopoMojo.Api/Features/User/ClaimsTransformation.cs

```diff
@@ -61,6 +48,116 @@ public async Task<ClaimsPrincipal> TransformAsync(ClaimsPrincipal principal)
+        ///     1. A typical subject-claim based user identification
+        ///     2. A claim that identifies the token as belong to a service account which matches an existing user's ServiceAccountClientId property.
+        /// 
+        /// The claim identifying service accounts will be consulted first if it is configured in Topo's OIDC settings and present in the claims principal. Otherwise,
+        /// this function falls back to the subject claim to match to a user.
+        /// </summary>
+        /// <param name="principal"></param>
+        /// <returns>A Topomojo User representing the result of the resolution process.</returns>
+        private async Task<User> ResolveUser(ClaimsPrincipal principal)
+        {
+            var subject = principal.Subject();
+            var serviceAccountClientId = ResolveServiceAccountClientId(principal);
+            var resolvedUser = default(User);
+
+            if (subject.IsEmpty() && !serviceAccountClientId.IsEmpty())
+            {
+                throw new ArgumentException($"""Can't resolve user: ClaimsPrincipal requires a "subject" claim or must have the claim {oidcOptions.AuthTypeClaimName} with value {oidcOptions.ServiceAccountAuthType}.""");
+            }
+
+            // first attempt to auth by service account, then by subject
+            if (serviceAccountClientId.NotEmpty())
+            {
+                if (!_cache.TryGetValue(serviceAccountClientId, out User serviceAccountUser))
+                {
+                    serviceAccountUser = await _svc.FindByServiceAccountClientId(serviceAccountClientId) ?? throw new Exception($"Service account client ID {serviceAccountClientId} didn't resolve to a user.");
+
+                    if (serviceAccountUser is null)
+                    {
+                        throw new UserServiceAccountResolutionFailed(serviceAccountClientId);
+                    }
+
+                    _cache.Set(serviceAccountClientId, serviceAccountUser, _cacheTimeout);
+                }
+
+                logger.LogInformation("Resolved user {userId} with service account ID {serviceAccountId}", serviceAccountUser.Id, serviceAccountClientId);
+                resolvedUser = serviceAccountUser;
+            }
+            else if (subject.NotEmpty())
+            {
+                if (!_cache.TryGetValue<User>(subject, out User user))
+                {
+                    user = await _svc.Load(subject) ?? new Models.User
+                    {
+                        Id = subject,
+                        Name = principal.Name()
+                    };
+
+                    // TODO: implement IChangeToken for this
+                    _cache.Set<User>(subject, user, _cacheTimeout);
+                }
+
+                logger.LogInformation("Resolved user {userId} via subject claim", user.Id);
+                resolvedUser = user;
+            }
+
+            if (resolvedUser is null)
```

Comment:
Feels like this be should comparing to default(User) or resolvedUser should be initalized to null.